### DISCUSSION TITLE
How do you construct complex filters like CONTAINS or IN?

### DISCUSSION THREAD

**Discussion by gi-stuart** [wopian/kitsu#985] [Question / Answer] [answered]
Would it be possible to add a more complex `filter` example to the docs, for example an `IN` condition?

I have tried the following and it only brings back a single item rather than an array:

```js
const { data } = await $jsonApi.get('/node/hubpage', {
        params: {
          filter: {
            drupal_internal__nid: {
              operator: 'IN',
                value: [1, 2]
            }
          }
        }
      })
```

**Comment by pedep** [comment, marked as answer]
The query in your example serializes to this:
```
params[filter][drupal_internal__nid][operator]=IN
&params[filter][drupal_internal__nid][value]=1
&params[filter][drupal_internal__nid][value]=2
```
which is not valid array syntax for most modern implementations.

Many deserializers will require the arrays to be suffixed with `[]` like this `params[filter][drupal_internal_nid][value][]=1`. This behaviour can be enabled in kitsu by providing the `modern` flag when creating your `Kitsu` instance https://github.com/wopian/kitsu/tree/master/packages/kitsu#parameters

The behaviour described above seems to be specifically discouraged in Drupal 
https://www.drupal.org/docs/core-modules-and-themes/core-modules/jsonapi-module/filtering#s-4-filtering-with-arrays-get-nodes-created-by-users-admin-john
>Tip: when utilizing square brackets for multiple values filters, do not just use empty square brackets for a new value.
While these work when typed into the URL, Guzzle and other http clients will only create one value as the array key will be seen to be same and override the previous value. It is better to use an index to create unique array elements.

To produce queries according to Drupals liking, you will have to format your queries like this

```
{
  params: {
    filter: {
      drupal_internal__nid: {
        operator: 'IN',
        value: {
          0: 1,
          1: 2
        }
      }
    }
  }
}
```

which produces output as expected by Drupal
```
params[filter][drupal_internal__nid][operator]=IN
&params[filter][drupal_internal__nid][value][0]=1
&params[filter][drupal_internal__nid][value][1]=2
```

**Reply by gi-stuart** [reply]
Thank you very much for taking the time to give such an excellent Drupal-specific explanation. The filter format you gave worked perfectly.

**Comment by wopian** [comment]
What is the URL query structure you are expecting to get sent to the API? Many of the server implementations implement complex filtering in vastly different ways so it is not clear what the result of the object you've provided is intended to be.

The library is fairly open ended to allow for these variances so you'll be able to construct the URL if we know what the API you are using expects the filter params to look like.

E.g one of the JSON:API servers I am currently working with expects complex filters like this: `playerpoints?filter=or(equals(rank,'17'),equals(rank,'18'))`, which can be passed as just a string, e.g:

```js
 client.get('playerpoints', {
   params: {
     filter: "or(equals(rank,'17'),equals(rank,'18'))"
     } })
```

**Reply by gi-stuart** [reply]
(Sorry for the delay in replying to pedep's response, I've been away and didn't get back to this until now).

The query structure I needed to send to Drupal was exactly as pedep noted above i.e.

```
params[filter][drupal_internal__nid][operator]=IN
&params[filter][drupal_internal__nid][value][0]=1
&params[filter][drupal_internal__nid][value][1]=2
```

Being able to provide a string for the filter is a nice option to have. Thank you!
